sql: Add information_schema.character_sets table

mknycha opened the pull request:
The `character_sets` table identifies the character sets available in the current database. Since CockroachDB supports only UTF-8 encoding at the moment ~and the collation is not stored per table, but column basis, every row in this table will look the same~, the encoding name in each row will be the same.
Release note (sql change): Added character_sets table to the information_schema.

Relates to: #8675

PostgreSQL table for reference: https://www.postgresql.org/docs/9.5/infoschema-character-sets.html

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56953)
<!-- Reviewable:end -->

otan commented on pkg/sql/information_schema.go:
```diff
@@ -262,6 +263,29 @@ https://www.postgresql.org/docs/9.5/infoschema-applicable-roles.html`,
+				tree.NewDString("UTF8"),    // character_set_name: UTF8 is the only available encoding
+				tree.NewDString("UCS"),     // character_repertoire: UCS for UTF8 encoding
+				tree.NewDString("UTF8"),    // form_of_use: same as the database encoding
+				tree.DNull,                 // default_collate_catalog
```
i think this one should be the current database name.

mknycha replied:
According to this [conversation](https://github.com/cockroachdb/cockroach/issues/8675#issuecomment-728349838), it should be `NULL`

mknycha replied:
Fixed

otan commented on pkg/sql/information_schema.go:
```diff
@@ -262,6 +263,29 @@ https://www.postgresql.org/docs/9.5/infoschema-applicable-roles.html`,
+				tree.NewDString("UCS"),     // character_repertoire: UCS for UTF8 encoding
+				tree.NewDString("UTF8"),    // form_of_use: same as the database encoding
+				tree.DNull,                 // default_collate_catalog
+				tree.DNull,                 // default_collate_schema
```
i think this should be `pg_catalog`

mknycha replied:
According to this [conversation](https://github.com/cockroachdb/cockroach/issues/8675#issuecomment-728349838), it should be `NULL`

mknycha replied:
Should not it be `pg_collation`?

otan replied:
looks like it should point to a valid collation. in this case, i think it's `pg_catalog`

otan commented on pkg/sql/information_schema.go:
```diff
@@ -262,6 +263,29 @@ https://www.postgresql.org/docs/9.5/infoschema-applicable-roles.html`,
+https://www.postgresql.org/docs/9.5/infoschema-character-sets.html`,
+	schema: vtable.InformationSchemaCharacterSets,
+	populate: func(ctx context.Context, p *planner, dbContext *dbdesc.Immutable, addRow func(...tree.Datum) error) error {
+		return forEachTableDesc(ctx, p, dbContext, hideVirtual, func(
```
there should only be one for each database in this instance, based on my comparison with postgres

mknycha replied:
Correct, fixed.

mknycha committed:
sql: Add information_schema.character_sets table

The `character_sets` table identifies the character sets available in
the current database. Since CockroachDB supports only UTF-8 encoding
at the moment, the encoding in each row is the same.
Release note (sql change): Added character_sets table
to the information_schema.

otan reviewed (commented):
ok, lgtm after fixing the test failures

otan commented:
i'll take a look at this and your other PR tomorrow. sorry for the delay.

mknycha commented:
That's totally ok, just wanted to fix the tests ;)
Please just note my comment in the `information_schema.go` regarding the `pg_catalog` table name, I am not sure about it.

otan commented on pkg/sql/information_schema.go:
```diff
@@ -262,6 +263,28 @@ https://www.postgresql.org/docs/9.5/infoschema-applicable-roles.html`,
+					tree.NewDString("UCS"),                    // character_repertoire: UCS for UTF8 encoding
+					tree.NewDString("UTF8"),                   // form_of_use: same as the database encoding
+					tree.NewDString(db.GetName()),             // default_collate_catalog
+					tree.DNull,                                // default_collate_schema
```
i think it's required we fill this in, let's make it `pg_catalog`.

mknycha replied:
Ok, fixed. I thought it was supposed to be the name of the table containing default collation, that's why I wanted to double-check.

otan commented on pkg/sql/information_schema.go:
```diff
@@ -262,6 +263,28 @@ https://www.postgresql.org/docs/9.5/infoschema-applicable-roles.html`,
+					tree.NewDString("UTF8"),                   // form_of_use: same as the database encoding
+					tree.NewDString(db.GetName()),             // default_collate_catalog
+					tree.DNull,                                // default_collate_schema
+					tree.NewDString(tree.DefaultCollationTag), // default_collate_name
```
looking at this again, this should actually be what our default collation is, i.e. `en-US`. sorry for the back and forth on it.

mknycha replied:
It's ok, fixed

otan reviewed (approved):
thanks! bors r+

otan commented:
bors r+

knz commented on pkg/sql/information_schema.go:
```diff
@@ -262,6 +263,28 @@ https://www.postgresql.org/docs/9.5/infoschema-applicable-roles.html`,
+					tree.NewDString("UCS"),        // character_repertoire: UCS for UTF8 encoding
+					tree.NewDString("UTF8"),       // form_of_use: same as the database encoding
+					tree.NewDString(db.GetName()), // default_collate_catalog
+					tree.NewDString("pg_catalog"), // default_collate_schema
```
regarding the last two columns - postgresql leaves those empty; are you sure that populating them will not crate an incompatibility?

otan replied:
do they?

```

 otan@otan-mbp  ~/go/src/github.com/cockroachdb/cockroach   srid_from_geojson  psql
psql (13.0)
Type "help" for help.

otan=# select * from information_schema.character_sets;
 character_set_catalog | character_set_schema | character_set_name | character_repertoire | form_of_use | default_collate_catalog | default_collate_schema | default_collate_name
-----------------------+----------------------+--------------------+----------------------+-------------+-------------------------+------------------------+----------------------
                       |                      | UTF8               | UCS                  | UTF8        | otan                    | pg_catalog             | ucs_basic
```

(happy to change this in a subsuquent PR if so)

knz replied:
I have this:
```
kena@kenax ~/cockroach % psql
psql (11.10)
Type "help" for help.

kena=> table information_schema.character_sets;
 character_set_catalog | character_set_schema | character_set_name | character_repertoire | form_of_use | default_collate_catalog | default_collate_schema | default_collate_name
-----------------------+----------------------+--------------------+----------------------+-------------+-------------------------+------------------------+----------------------
                       |                      | UTF8               | UCS                  | UTF8        | kena                    |                        |
(1 row)
```

otan commented:
bors r-

looks like they do not in earlier versions of postgres, let me check it out.

knz commented:
Also even if you keep the columns populated, don't we need `ucs_basic` in the last instead of `en-US`? What's the constraint there?

otan commented:
we don't support `ucs_basic` as a collation type, we default to `en-US`.

otan commented:
According to the postgres docs: 

`Name of the default collation. The default collation is identified as the collation that matches the COLLATE and CTYPE settings of the current database. If there is no such collation, then this column and the associated schema and catalog columns are null.`

something must've changed in PG 13 for this to be ucs_basic as default by now, but since we're aiming for PG 9/10 i think we can use NULL. i'll make these changes for you @mknycha, your name will still be on the commit.

otan commented:
bors r+

Amruta-Ranade commented:
@mknycha Thank you for contributing to CockroachDB this year. As a token of our appreciation, we would like to send you a gift. Please DM me on our [community Slack](https://cockroa.ch/slack) @Amruta so I can send you a link. (If you don’t want a gift, we also have a charitable contribution choice.) All orders need to be in by December 13, so please contact me as soon as possible :)
